src/python/pybind11/src/pyxrt.cpp

```diff
@@ -71,95 +79,128 @@ py::class_<xrt::uuid>(m, "uuid")
-    .def("read_register", &xrt::kernel::read_register)
-    ;
+    .def("read_register", &xrt::kernel::read_register);
 
 
 /*
  *
  * XRT:: BO
  *
  */
-py::class_<xrt::bo>(m, "bo")
-    .def(py::init<xrt::device,size_t,xrt::buffer_flags,xrt::memory_group>())
+py::class_<xrt::bo> pybo(m, "bo");
+
+py::enum_<xrt::bo::flags>(pybo, "")
+    .value("normal", xrt::bo::flags::normal)
+    .value("cacheable", xrt::bo::flags::cacheable)
+    .value("device_only", xrt::bo::flags::device_only)
+    .value("host_only", xrt::bo::flags::host_only)
+    .value("p2p", xrt::bo::flags::p2p)
+    .value("svm", xrt::bo::flags::svm)
+    .export_values();
+
+pybo.def(py::init<xrt::device, size_t, xrt::bo::flags, xrt::memory_group>())
     .def(py::init<xrt::bo, size_t, size_t>())
     .def("write", ([](xrt::bo &b, py::buffer pyb, size_t seek)  {
-	  py::buffer_info info = pyb.request();
-	  b.write(info.ptr, info.itemsize * info.size , seek);
-    }))
+                       py::buffer_info info = pyb.request();
+                       b.write(info.ptr, info.itemsize * info.size , seek);
+                   }))
     .def("read", ([](xrt::bo &b, size_t size, size_t skip) {
-	  int nitems = size/sizeof(int);
-	  py::array_t<int> result = py::array_t<int>(nitems);
-	  
-	  py::buffer_info bufinfo = result.request();
-	  int* bufptr = (int*) bufinfo.ptr;
-	  b.read(bufptr, size, skip);
-	  return result;
-     }))
-  .def("sync", ([](xrt::bo &b, xclBOSyncDirection dir, size_t size, size_t offset)  {	
-	b.sync(dir, size, offset);
-      }))
-  .def("size", &xrt::bo::size)
-  .def("address", &xrt::bo::address)
+                      int nitems = size/sizeof(int);
+                      py::array_t<int> result = py::array_t<int>(nitems);
+
+                      py::buffer_info bufinfo = result.request();
+                      int* bufptr = (int*) bufinfo.ptr;
+                      b.read(bufptr, size, skip);
+                      return result;
+                  }))
+    .def("sync", ([](xrt::bo &b, xclBOSyncDirection dir, size_t size, size_t offset)  {
+                      b.sync(dir, size, offset);
+                  }))
+    .def("map", ([](xrt::bo &b)  {
+                     /*
+                      * memoryview ctor has a bug where it uses c_str() of info.format directly
+                      * which leads to stale pointer since info goes out of scope. See the
+                      * following code snipet from memoryview contructor in ptypes.h:
+                      * buf.format = const_cast<char *>(info.format.c_str());
+                      *
+                      * To resolve this we are creating buffer_info as static object so the
+                      * format string is persistent. We reuse info object by updating ptr and
+                      * length for each new bo.
+                      */
+                     static py::buffer_info info(b.map(),
+                                          sizeof(unsigned int),
```

Comment:
are all possible buffer sizes a multiple of sizeof(uint)?  Or is char safer?

Reply:
Migrated to native py::memoryview::from_memory() available in 2.6.0 which does not have the bugs found before. It natively uses char as datatype.